src/pymoca/tree.py

```diff
@@ -259,6 +273,694 @@ def handle_walk(self, listener: TreeListener, tree: Union[ast.Node, dict, list])
+    7. Repeat 1-6 for each lexically enclosing instance scope, stopping at `encapsulated`
+    unless predefined type, function, operator. If name matches a variable (a.k.a. component,
+    symbol) in an enclosing class, it must be a `constant`.
+    """
+
+    current_scope = scope
+    while True:
+        if (
+            (
+                found := _find_local(
+                    name,
+                    current_scope,
+                    check_builtin_classes=check_builtin_classes,
+                )
+            )
+            or (
+                found := _find_inherited(
+                    name,
+                    current_scope,
+                    check_builtin_classes=check_builtin_classes,
+                    current_extends=current_extends,
+                )
+            )
+            or search_imports
+            and (
+                found := _find_imported(
+                    name,
+                    current_scope,
+                    check_builtin_classes=check_builtin_classes,
+                    current_extends=current_extends,
+                )
+            )
+            or not search_parent
+            or not current_scope.parent
+            or current_scope.encapsulated
+        ):
+            break
+        current_scope = current_scope.parent
+    # If name matches a variable (a.k.a. component a.k.a. symbol) in an enclosing class,
+    # it must be a `constant`.
+    if (
+        isinstance(found, ast.Symbol)
+        and current_scope != scope
+        and "constant" not in found.prefixes
+    ):
+        raise NameLookupError("Non-constant Symbol found in enclosing class")
+
+    # If not found and we stopped at an encapsulated class,
+    # then search predefined functions and operators in global scope
+    # TODO: Add predefined functions and operators to global scope before this
+    if found is None and current_scope.encapsulated:
+        found = _find_local(name, scope.root)
+        if not isinstance(found, ast.Class) or found.type not in ("function", "operator"):
+            found = None
+
+    return found
+
+
+def _find_rest_of_name(
+    first: Union[ast.Class, ast.Symbol], rest_of_name: str
+) -> Optional[Union[ast.Class, ast.Symbol]]:
+    """Lookup the `B.C` part of Composite Name Lookup (`A.B.C`) (spec 5.3.2)
+
+    1. `A` is looked up using Simple Name Lookup and passed as `first` argument
+    2. If `A` is a Component:
+        1. `B.C` is looked up from named component elements of `A`
+        2. if not found and if `A.B.C` is used as a function call and `A` is a scalar or can be
+        evaluated as a scalar from an array and `B` and `C` are classes,
+        it is a non-operator function call.
+    3. If `A` is a Class:
+        1. `A` is temporarily flattened without modifiers
+        2. `B.C` is looked up among named elements of temp flattened class,
+        but if `A` is not a package, lookup is restricted to `encapsulated` elements only
```

Comment:
Feel free to look into it, but I think I can answer your questions.

> * Is it only "stylistic", in the sense that removing the keyword would mean functionally the same model if it was correct? Or will lookup actually change in some cases?

The answer is yes lookup will change. Both simple and composite name lookup are affected by `encapsulated`. Take it out and they happen differently.

> * Examples / test cases in case the latter of the above (lookup changes) is true.
> * Is there cases in which we can avoid naming conflicts by using that classifier, and if so, do we have a test case for that?

There are several cases in name_lookup_test.py that cover `encapsulated` and I believe they cover these cases.